// @angular/cdk/esm5/accordion/accordion-module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CdkAccordion } from './accordion';
import { CdkAccordionItem } from './accordion-item';
import * as ɵngcc0 from '@angular/core';
var CdkAccordionModule = /** @class */ (function () {
    function CdkAccordionModule() {
    }
CdkAccordionModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CdkAccordionModule });
CdkAccordionModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function CdkAccordionModule_Factory(t) { return new (t || CdkAccordionModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CdkAccordionModule, { declarations: function () { return [CdkAccordion, CdkAccordionItem]; }, exports: function () { return [CdkAccordion, CdkAccordionItem]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CdkAccordionModule, [{
        type: NgModule,
        args: [{
                exports: [CdkAccordion, CdkAccordionItem],
                declarations: [CdkAccordion, CdkAccordionItem]
            }]
    }], function () { return []; }, null); })();
    return CdkAccordionModule;
}());
export { CdkAccordionModule };
